server_http.go

```diff
@@ -0,0 +1,220 @@
+func NewHttpServer(iface string, searcher Searcher) *HttpServer {
+	return &HttpServer{
+		iface:    iface,
+		Searcher: searcher,
+		Logger:   log.New(os.Stderr, "[httpserver] ", log.LstdFlags),
+	}
+}
+
+// Start instructs the Server to bind to the interface and accept connections.
+func (s *HttpServer) Start() error {
+
+	ln, err := net.Listen("tcp", s.iface)
+	if err != nil {
+		return err
+	}
+
+	s.addr = ln.Addr()
+
+	s.template, err = template.New("ServerTemplate").Parse(templateSource)
+	if err != nil {
+		ln.Close()
+		return err
+	}
+
+	CSRF := csrf.Protect(securecookie.GenerateRandomKey(32),
+		csrf.Secure(false),
+		csrf.HttpOnly(true))
+
+	go http.Serve(ln, CSRF(s))
+	return nil
+}
+
+// Addr returns the address to which the Server is bound.
+func (s *HttpServer) Addr() net.Addr {
+	return s.addr
+}
+
+// ServeHTTP implements a http.Handler, serving the query interface for Ekanite
+func (s *HttpServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	dontCache(w, r)
+
+	if r.Method == "GET" || r.Method == "HEAD" {
+		//HEAD is conveniently supported by net/http without further action
+		serveIndex(s, w, r)
+
+	} else if r.Method == "POST" {
+
+		err := r.ParseForm()
+		if err != nil {
+			s.Logger.Printf("Error parsing form '%s'", err)
+			http.Error(w, "Error parsing form", http.StatusBadRequest)
+			return
+		}
+
+		if len(r.FormValue("query")) == 0 {
+			serveIndex(s, w, r)
+			return
+		}
+
+		userQuery := r.FormValue("query")
+
+		s.Logger.Printf("executing query '%s'", userQuery)
+
+		resultSet, err := s.Searcher.Search(userQuery)
+		var resultSlice []string
+
+		if err != nil {
+			s.Logger.Printf("Error executing query: '%s'", err)
+			http.Error(w, "Error executing query: "+err.Error(), http.StatusInternalServerError)
+			return
+		} else {
+			for s := range resultSet {
+				resultSlice = append(resultSlice, s)
+			}
+		}
+
+		data := struct {
+			Title         string
+			Headline      string
+			ReturnResults bool
+			LogMessages   []string
+			CsrfField     template.HTML
+		}{
+			"Ekanite query interface",
+			"Ekanite - Listing " + strconv.Itoa(len(resultSlice)) + " results for '" + userQuery + "'",
+			true,
+			resultSlice,
+			csrf.TemplateField(r),
+		}
+
+		if err := s.template.Execute(w, data); err != nil {
+			s.Logger.Print("Error executing template: ", err)
+		}
+
+	} else {
+		http.Error(w, "Unsupported method", http.StatusMethodNotAllowed)
+	}
+}
+
+// serveIndex serves the plain index for the GET request and POST failovers
+func serveIndex(s *HttpServer, w http.ResponseWriter, r *http.Request) {
+	data := struct {
+		Title         string
+		Headline      string
+		ReturnResults bool
+		LogMessages   []string
+		CsrfField     template.HTML
+	}{
+		"Ekanite query interface",
+		"Ekanite query interface",
+		false,
+		[]string{},
+		csrf.TemplateField(r),
+	}
+
+	if err := s.template.Execute(w, data); err != nil {
+		s.Logger.Print("Error executing template: ", err)
```

Comment:
I never saw an error from template.Execute in production, provided the correct struct is passed in. Nonetheless I will implement your suggestion and let ```serveIndex``` return an error to the caller.